### PR TITLE
Cache the stake pool metadata registry

### PR DESCRIPTION
**Pull request by rvl**
Relates to #1164.

# Overview

Caches the downloaded stake pool metadata registry zip file in a temporary directory.

### CODE REVIEW

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 1]
```diff
@@ -194,31 +198,59 @@ defaultRecordTypeOptions = Aeson.defaultOptions
...
-    req <- parseUrlThrow url
-    httpSink req $ \_response -> Conduit.mapM_ (BS.hPut zipFile)
-    hTell zipFile
+    checkCached = try (getModificationTime zipFileName) >>= \case
+        Right mtime -> do
+            now <- getCurrentTime
+            pure $ if (diffUTCTime now mtime < cacheTTL)
+                then Just mtime
+                else Nothing
+        Left (_ :: IOException) -> pure Nothing
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Metadata.hs]: **[question]**

I suppose that `getModificationTime` throws an IOException when the file doesn't exist?

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 2]
```diff
@@ -377,3 +411,5 @@ instance ToText RegistryLogMsg where
             fmt $ "Could not parse metadata: "+|e|+""
         MsgExtractFileResult (Just (Right _)) ->
             "Successfully parsed metadata"
+        MsgUsingCached _ mtime ->
+            "Using cached file last modified at " <> T.pack (show mtime)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Metadata.hs:433]: :+1:

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 3]
```diff
@@ -194,31 +198,59 @@ defaultRecordTypeOptions = Aeson.defaultOptions
...
-        fetchStakePoolMetaZip url zipFile >>= \case
-            Right size -> do
-                logTrace tr' $ MsgDownloadComplete size
-                hClose zipFile
-                Right <$> getMetadataFromZip tr' zipFileName poolOwners
-            Left e -> pure $ Left e
+getStakePoolMetadata tr cacheDir url poolOwners = do
+    let zipFileName = cacheDir </> "registry.zip"
+    let tr' = contramap (fmap (RegistryLog url zipFileName)) tr
+    createDirectoryIfMissing False cacheDir
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Metadata.hs]: **[question]**

Why not also create parent directory (i.e. ~~False~~ -> True) if they don't exist to easily allow creating nested directories? :thinking:

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 4]
```diff
@@ -194,31 +198,59 @@ defaultRecordTypeOptions = Aeson.defaultOptions
...
+    fetchStakePoolMetaZipCached tr' url zipFileName >>= \case
+        Right _ -> Right <$> getMetadataFromZip tr' zipFileName poolOwners
+        Left e -> pure $ Left e
+
+cacheTTL :: NominalDiffTime
+cacheTTL = 3600
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Metadata.hs]: **[suggestion]**

Add a comment about the duration. `NominalDiffTime` interprets `Int` as seconds for conversions through the `Num` instance, but that's not necessarily obvious :sweat_smile:

File: lib/core/test/unit/Cardano/Pool/MetadataSpec.hs [hunk 5]
```diff
@@ -76,26 +75,47 @@ import qualified Data.Text as T
...
 
+        it "Caches the zip" $ \(baseUrl, metadataDir, tr, getMsgs) -> do
+            void $ getStakePoolMetadata tr metadataDir (testUrl baseUrl) presentOwners
+
+            res <- getStakePoolMetadata tr metadataDir (testUrl baseUrl) presentOwners
+            res `shouldBe` Right (map Just presentMetas)
+
+            msgs <- map registryLogMsg <$> getMsgs
+            length (filter isMsgDownloadStarted msgs) `shouldBe` 1
+            length (filter isMsgUsingCached msgs) `shouldBe` 1
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Pool/MetadataSpec.hs:105]: **[remark]**

It's kinda cool to be able to test log messages directly.

File: lib/core/src/Cardano/Pool/Metadata.hs (hunk 3 above)
Comment (rvl) [reply lib/core/src/Cardano/Pool/Metadata.hs]: Yeah, maybe. I was originally planning to bury this file in the database directory, which should exist. But actually the database is optional, so now it's now a temporary directory. So to resolve this question I now remove the createDirectoryIfMissing altogether.

File: lib/core/test/unit/Cardano/Pool/MetadataSpec.hs (hunk 5 above)
Comment (rvl) [reply lib/core/test/unit/Cardano/Pool/MetadataSpec.hs:105]: Yes, Neil told me this also. The tests can easily find out what your program is doing.

File: lib/core/test/unit/Cardano/Pool/MetadataSpec.hs [hunk 6]
```diff
@@ -76,26 +75,47 @@ import qualified Data.Text as T
...
+
+        it "Refreshes expired cache" $ \(baseUrl, metadataDir, tr, getMsgs) -> do
+            void $ getStakePoolMetadata tr metadataDir (testUrl baseUrl) presentOwners
+
+            old <- makeExpired <$> getCurrentTime
+            forFilesIn metadataDir $ \f -> setModificationTime f old
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Pool/MetadataSpec.hs]: **[suggestion]**

Wouldn't be "simpler" to give the cache TTL to `getStakePoolMetadata` as an argument. We could make use of a config record type such as:

```hs
data MetadataConfig = MetadataConfig
    { cacheArchive :: FilePath
    , cacheTTL :: NominalDiffTime
    , registryURL :: String
    }
```

With a default configuration and an accessor like `getMetadataConfig :: IO MetadataConfig` which takes care of overriding the URL with the right ENV var.

Thoughts :thinking: ?

**Review by KtorZ — APPROVED** [review]

File: lib/core/test/unit/Cardano/Pool/MetadataSpec.hs (hunk 6 above)
Comment (rvl) [reply lib/core/test/unit/Cardano/Pool/MetadataSpec.hs]: Yes, good suggestion.

**Comment by rvl** [comment]
OK I have addressed the review comments.

**Review by KtorZ — APPROVED** [review]

**Pull request merged**